### ISSUE TITLE
Thread not killed after calling #kill

### ISSUE THREAD

**Issue by wjordan** [jruby/jruby#6188]
This simple script demonstrates an edge-case where `Thread#kill` does not kill a thread as expected on JRuby. It intermittently hangs on JRuby but reliably runs on other Ruby implementations (MRI, TruffleRuby):

```ruby
Thread.new {sleep rescue sleep}.
  tap {|t| t.raise 'Error'}.
  tap(&:kill).
  tap(&:join)
```

Some notes:
- It seems to be a race condition / timing issue related to the rescue operation?
- If I add a `sleep 0.1` after `t.raise 'Error'`, the script passes
- If I change `t.raise 'Error'` to `t.raise RuntimeError.new` (or `t.raise`), the script passes

**Environment Information**
JRuby version 9.2.11.1 on Ubuntu 18.04
```
jruby 9.2.11.1 (2.5.7) 2020-03-25 b1f55b1a40 Java HotSpot(TM) 64-Bit Server VM 25.131-b11 on 1.8.0_131-b11 [linux-x86_64]
Linux wjordan 4.15.0-88-generic #88-Ubuntu SMP Tue Feb 11 20:11:34 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by ivoanjo** [comment]
I'm just a lurker, but I was curious about this one, so I took a stab at debugging it.

Out of the box, I can reproduce this issue with the latest JRuby trunk (`jruby 9.3.0.0-SNAPSHOT (2.6.5) 2020-04-23 d6fa59fde4 OpenJDK 64-Bit Server VM 14.0.1+7 on 14.0.1+7 [linux-x86_64]`) if I execute JRuby in a loop until it gets stuck. Usually it takes 10-20 tries, but sometimes a bit more.

I'm pretty sure this is a race around the interrupt handling code. Both `raise` and `kill` set the interrupt status on the other thread, and what I observe when the example code gets stuck is that the thread:

1. Rescues the exception
2. Goes inside the second sleep
3. Has no interrupt flags
4. Still has the unprocessed kill interrupt inside the `pendingInterruptQueue`

Furthermore, if I add any kind of delay inside `pollThreadEvents` from `RubyThread.java`

```java
    // CHECK_INTS
    public void pollThreadEvents(ThreadContext context) {
        if (anyInterrupted()) {
            executeInterrupts(context, true);
        }
    }
```

e.g. something like this

```java
    // CHECK_INTS
    public void pollThreadEvents(ThreadContext context) {
        if (anyInterrupted()) {
            try {
                Thread.sleep(100);
            } catch (Exception e) {
                e.printStackTrace();
            }
            executeInterrupts(context, true);
        }
    }
```

or even a `println`, then suddendly I can reproduce this 100% of the time. So this leads me to believe that the issue is that if the "interrupted" thread gets interrupted twice, but resets back the interrupted state without noticing that there's actually two interrupts.

I ran out of time for today, so hopefully this is helpful. Hopefully I can come back to it soon and maybe provide a patch :)

**Comment by headius** [comment]
@ivoanjo Good investigation! Seems likely that this needs to be made a bit more robust when there's multiple events getting tossed around.

**Comment by headius** [comment]
Link with #6569.

**Comment by headius** [comment]
This may have been fixed by my `handle_interrupt` changes in #6569 because it no longer hangs on master.

**Comment by headius** [comment]
I can't reproduce anymore! Going to close this as fixed, probably due to #6569.
